### ISSUE TITLE
Floating versions not supported since version 3.0.0

### ISSUE THREAD

**Issue by Droppers** [CycloneDX/cyclonedx-dotnet#814]
I cannot upgrade to the latest version of dotnet CycloneDX because it does not support floating versions anymore. Has the support for this been removed on purpose or accidentally, since I do not see this in the breaking changes in version 3.0.0?

Take the following package reference as an example:
```xml
<PackageReference Include="JetBrains.Annotations" Version="2023.*" />
```

Which results in the following stack trace:

```
Unhandled exception: System.ArgumentException: '2023.*' is not a valid version string. (Parameter 'value')
   at NuGet.Versioning.NuGetVersion.Parse(String value)
   at NuGet.Versioning.NuGetVersion..ctor(String version)
   at CycloneDX.Services.ProjectAssetsFileService.<>c.<GetDotnetDependencys>b__3_2(String d) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/ProjectAssetsFileService.cs:line 63
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.Any[TSource](IEnumerable`1 source, Func`2 predicate)
   at CycloneDX.Services.ProjectAssetsFileService.GetDotnetDependencys(String projectFilePath, String projectAssetsFilePath, Boolean isTestProject) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/ProjectAssetsFileService.cs:line 72
   at CycloneDX.Services.ProjectFileService.GetProjectDotnetDependencysAsync(String projectFilePath, String baseIntermediateOutputPath, Boolean excludeTestProjects, String framework, String runtime) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/ProjectFileService.cs:line 186
   at CycloneDX.Services.SolutionFileService.GetSolutionDotnetDependencys(String solutionFilePath, String baseIntermediateOutputPath, Boolean excludeTestProjects, String framework, String runtime) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Services/SolutionFileService.cs:line 116
   at CycloneDX.Runner.HandleCommandAsync(RunOptions options) in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Runner.cs:line 167
   at CycloneDX.Program.<>c__DisplayClass0_0.<<Main>b__2>d.MoveNext() in /home/runner/work/cyclonedx-dotnet/cyclonedx-dotnet/CycloneDX/Program.cs:line 137
```

https://learn.microsoft.com/en-us/nuget/concepts/package-versioning#floating-version-resolutions

**Comment by mtsfoni** [comment]
Hi droppers,

thanks for reporting the issue.

To me, this looks like a bug caused by a new feature. I will check it and release a patch version. Hopefully later today, but the bug might be a little tricky.

I will be honest: I didn't even know you could do that. I'll certainly add one or two testcases for floating versions after the fix.

**Comment by shaeussler** [comment]
Thanks for the fast fix.

**Comment by mtsfoni** [comment]
Providing some context for documentation:

The new feature -ipr needed a new way to figure out which dependencies are "direct" (will hang on the metacomponent in the dependency graph).  This new way utilizes the `projectFileDependencyGroups` of the assets file, where the version has the floating representation. 

To figure out, which dependency is direct, a comparison was made between name and version of the actual library and the entry in the `projectFileDependencyGroups`. For that, the version in that group that to be converted to a NuGetVersion (class in NuGet library), which doesn't support floating versions, thus the invalid string exception.

As a fix, I removed the comparison of version and go only by the name, which should probably be sufficient, as to my knowledge it's not possible to have two different libraries with the same name.

The bug should be fixed in version 3.0.4, regression-test exists now.

Please let me know if you have further problems.

**Comment by Droppers** [comment]
Thank you for quickly resolving the bug 👍🏻 !
